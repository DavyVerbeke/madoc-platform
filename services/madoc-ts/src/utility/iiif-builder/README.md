# IIIF Builder
An extension to Hyperion framework for building and editing IIIF resources in a vault store.


## Creating new manifests
```js
const builder = new IIIFBuilder();

builder.createManifest('http://example.org/manifest-1', manifest => {
  manifest.addLabel('My manifest', 'en');
  
  manifest.summary = { en: ['Summary of this manifest'] };
  
  
});
```


### Creating Presentation 3.0 MVM
Source: https://iiif.io/api/cookbook/recipe/0001-mvm-image/manifest.json
```js

const builder = new IIIFBuilder();

builder.createManifest(
  'https://iiif.io/api/cookbook/recipe/0001-mvm-image/manifest.json', 
  manifest => {
    manifest.addLabel('Image 1', 'en');
    manifest.createCanvas('https://iiif.io/api/cookbook/recipe/0001-mvm-image/canvas/p1', canvas => {
      canvas.width = 1800;
      canvas.height = 1200;
      canvas.createAnnotation(
        'https://iiif.io/api/cookbook/recipe/0001-mvm-image/annotation/p0001-image',
        {
          id: 'https://iiif.io/api/cookbook/recipe/0001-mvm-image/annotation/p0001-image',
          type: 'Annotation',
          motivation: 'painting',
          body: {
            id: 'http://iiif.io/api/presentation/2.1/example/fixtures/resources/page1-full.png',
            type: 'Image',
            format: 'image/png',
            height: 1800,
            width: 1200
          }
        }
      );
    })
  }
);
```

### Serialising
You always work with IIIF Presentation 3 with the builder.

```js
const builder = new IIIFBuilder();

const myManifest = builder.createManifest('http://example.org/manifest-1', manifest => {
  manifest.addLabel('My manifest', 'en');
});

builder.toPresentation2(myManifest); // presentation 2 manifest
builder.toPresentation3(myManifest); // presentation 3 manifest

```

## Editing resources

You can pass in your own Vault instance and edit resources in your vault. When changes
are made to resources any subscribers to the vault (such as React) will be notified, 
allowing UI to be kept in sync.

```js

const vault = new Vault();
const builder = new IIIFBuilder(vault);

// Load from somewhere
const manifestRef = await vault.loadManifest('http://example.org/manifest-1.json');

builder.editManifest('http://example.org/manifest-1.json', manifest => {
  manifest.setLabel({ en: ['Some different label'] });
});

// Elsewhere in your code (usually handled by framework abstraction)
vault.subscribe(
  s => s.hyperion.entities.Manifest['http://example.org/manifest-1.json'].label, 
  newLabel => {
    console.log('Label changed => ', newLabel)
  }
);
```

## Framework integrations

**Still a work in progress**

```js

function ManifestLabelEditor() {
  const manifest = useManifest();
  const editor = useEditor(manifest);
  const [value, setValue] = useState(manifest.label.en[0]);
  
  return <>
    <input 
      type="text"
      value={value}
      onChange={e => setValue(e.target.value)} 
    />
    <button onClick={() => editor.setLabel({ en: [value] })}>
      save changes
    </button>
   </>;
}

```
