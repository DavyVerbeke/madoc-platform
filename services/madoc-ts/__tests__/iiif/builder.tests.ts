import { Vault } from '@hyperion-framework/vault';
import { entityActions } from '@hyperion-framework/store';
import { emptyManifest } from '@hyperion-framework/parser';
import { IIIFBuilder } from '../../src/utility/iiif-builder/iiif-builder';

describe('iiif vault building', () => {
  const vault = new Vault();
  const store = vault.getStore();

  test('You can create new IIIF manifest', () => {
    store.dispatch(
      entityActions.importEntities({
        entities: {
          Manifest: {
            'http://example.org/manifest-1': { ...emptyManifest },
          },
          Annotation: {},
          AnnotationCollection: {},
          AnnotationPage: {},
          Canvas: {},
          Collection: {},
          ContentResource: {},
          Range: {},
          Selector: {},
          Service: {},
        },
      })
    );

    expect(store.getState().hyperion.entities.Manifest).toMatchInlineSnapshot(`
      Object {
        "http://example.org/manifest-1": Object {
          "accompanyingCanvas": null,
          "annotations": Array [],
          "behavior": Array [],
          "homepage": null,
          "id": "https://hyperion/empty-manifest",
          "items": Array [],
          "label": null,
          "logo": Array [],
          "metadata": Array [],
          "motivation": null,
          "navDate": null,
          "partOf": Array [],
          "placeholderCanvas": null,
          "posterCanvas": null,
          "provider": Array [],
          "rendering": Array [],
          "requiredStatement": null,
          "rights": null,
          "seeAlso": Array [],
          "service": Array [],
          "services": Array [],
          "start": null,
          "structures": Array [],
          "summary": null,
          "thumbnail": Array [],
          "type": "Manifest",
          "viewingDirection": "left-to-right",
        },
      }
    `);
  });

  test('You can add fields', () => {
    store.dispatch(
      entityActions.modifyEntityField({
        id: 'http://example.org/manifest-1',
        type: 'Manifest',
        value: { none: ['An example label'] },
        key: 'label',
      })
    );

    expect(store.getState().hyperion.entities.Manifest).toMatchInlineSnapshot(`
      Object {
        "http://example.org/manifest-1": Object {
          "accompanyingCanvas": null,
          "annotations": Array [],
          "behavior": Array [],
          "homepage": null,
          "id": "https://hyperion/empty-manifest",
          "items": Array [],
          "label": Object {
            "none": Array [
              "An example label",
            ],
          },
          "logo": Array [],
          "metadata": Array [],
          "motivation": null,
          "navDate": null,
          "partOf": Array [],
          "placeholderCanvas": null,
          "posterCanvas": null,
          "provider": Array [],
          "rendering": Array [],
          "requiredStatement": null,
          "rights": null,
          "seeAlso": Array [],
          "service": Array [],
          "services": Array [],
          "start": null,
          "structures": Array [],
          "summary": null,
          "thumbnail": Array [],
          "type": "Manifest",
          "viewingDirection": "left-to-right",
        },
      }
    `);
  });

  test('builder class', () => {
    const builder = new IIIFBuilder();

    const newManifest = builder.createManifest('http://example.org/manifest-1', manifest => {
      manifest.addLabel('Testing a label');
    });

    expect(newManifest.label).toMatchInlineSnapshot(`
      Object {
        "none": Array [
          "Testing a label",
        ],
      }
    `);

    const editedManifest = builder.editManifest('http://example.org/manifest-1', manifest => {
      manifest.label = { en: ['A new label'] };
    });

    expect(editedManifest.label).toMatchInlineSnapshot(`
      Object {
        "en": Array [
          "A new label",
        ],
      }
    `);
  });

  test('building manifest with canvas', () => {
    const builder = new IIIFBuilder();

    builder.createManifest('http://example.org/manifest', manifest => {
      manifest.addLabel('My manifest');

      manifest.createCanvas('http://example.org/manifest/canvas-1', canvas => {
        canvas.addLabel('Canvas 1');
      });

      manifest.createCanvas('http://example.org/manifest/canvas-2', canvas => {
        canvas.addLabel('Canvas 1');
      });
    });

    expect(builder.vault.getState().hyperion).toMatchSnapshot();
  });
});
