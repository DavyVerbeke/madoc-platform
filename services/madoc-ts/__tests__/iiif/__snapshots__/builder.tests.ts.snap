// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iiif vault building building manifest with canvas 1`] = `
Object {
  "entities": Object {
    "Annotation": Object {},
    "AnnotationCollection": Object {},
    "AnnotationPage": Object {},
    "Canvas": Object {
      "http://example.org/manifest/canvas-1": Object {
        "accompanyingCanvas": null,
        "annotations": Array [],
        "behavior": Array [],
        "duration": 0,
        "height": 0,
        "homepage": null,
        "id": "http://example.org/manifest/canvas-1",
        "items": Array [],
        "label": Object {
          "none": Array [
            "Canvas 1",
          ],
        },
        "logo": Array [],
        "metadata": Array [],
        "motivation": null,
        "navDate": null,
        "partOf": Array [],
        "placeholderCanvas": null,
        "posterCanvas": null,
        "provider": Array [],
        "rendering": Array [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": Array [],
        "service": Array [],
        "summary": null,
        "thumbnail": Array [],
        "type": "Canvas",
        "width": 0,
      },
      "http://example.org/manifest/canvas-2": Object {
        "accompanyingCanvas": null,
        "annotations": Array [],
        "behavior": Array [],
        "duration": 0,
        "height": 0,
        "homepage": null,
        "id": "http://example.org/manifest/canvas-2",
        "items": Array [],
        "label": Object {
          "none": Array [
            "Canvas 1",
          ],
        },
        "logo": Array [],
        "metadata": Array [],
        "motivation": null,
        "navDate": null,
        "partOf": Array [],
        "placeholderCanvas": null,
        "posterCanvas": null,
        "provider": Array [],
        "rendering": Array [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": Array [],
        "service": Array [],
        "summary": null,
        "thumbnail": Array [],
        "type": "Canvas",
        "width": 0,
      },
    },
    "Collection": Object {},
    "ContentResource": Object {},
    "Manifest": Object {
      "http://example.org/manifest": Object {
        "accompanyingCanvas": null,
        "annotations": Array [],
        "behavior": Array [],
        "homepage": null,
        "id": "http://example.org/manifest",
        "items": Array [
          Object {
            "id": "http://example.org/manifest/canvas-1",
            "type": "Canvas",
          },
          Object {
            "id": "http://example.org/manifest/canvas-2",
            "type": "Canvas",
          },
        ],
        "label": Object {
          "none": Array [
            "My manifest",
          ],
        },
        "logo": Array [],
        "metadata": Array [],
        "motivation": null,
        "navDate": null,
        "partOf": Array [],
        "placeholderCanvas": null,
        "posterCanvas": null,
        "provider": Array [],
        "rendering": Array [],
        "requiredStatement": null,
        "rights": null,
        "seeAlso": Array [],
        "service": Array [],
        "services": Array [],
        "start": null,
        "structures": Array [],
        "summary": null,
        "thumbnail": Array [],
        "type": "Manifest",
        "viewingDirection": "left-to-right",
      },
    },
    "Range": Object {},
    "Selector": Object {},
    "Service": Object {},
  },
  "mapping": Object {
    "http://example.org/manifest": "Manifest",
    "http://example.org/manifest/canvas-1": "Canvas",
    "http://example.org/manifest/canvas-2": "Canvas",
  },
  "requests": Object {},
}
`;
